app/src/main/java/sun/trainingcourse/appcaculator/MainActivity.kt

```diff
@@ -1,12 +1,221 @@
+                if (!isOperator(c)) {
+                    appendEditTextMath(")", true)
+                }
+            }
+        }
+
+        btn_clear.setOnClickListener {
+            txt_value.text = ""
+            edt_math.text.clear()
+        }
+
+        btn_delete.setOnClickListener {
+            val txt = BaseInputConnection(edt_math, true)
+            txt.sendKeyEvent(KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_DEL))
+        }
+
+        btn_equal.setOnClickListener {
+            val value = edt_math.text.toString()
+            if (value.isNotEmpty()) {
+                val c = value.substring(value.length - 1, value.length)[0]
+                if (!isOperator(c) || c == ')') {
+                    val text = edt_math.text.toString()
+                    val array = postfix(processString(text))
+                    txt_value.text = valueMath(array)
+                }
+            }
+        }
+    }
+
+    private fun isOperator(c: Char): Boolean {
+        return when (c) {
+            '+', '-', '*', '/', '(', ')' -> true
+            else -> false
+        }
+    }
+
+    private fun priority(c: Char): Int {
+        return if (c == '+' || c == '-') 1
+        else if (c == '*' || c == '/') 2
+        else 0
+    }
+
+    private fun processString(text: String): Array<String> {
+        var math = text
+        var s = ""
+        val elementMath: Array<String>
+        math = math.trim()
+        math = math.replace("\\s+".toRegex(), " ")
+        for (element in math) {
+            s = if (!isOperator(element)) s + element else "$s $element "
+        }
+        s = s.trim()
+        s = s.replace("\\s+".toRegex(), " ")
+        elementMath = s.split(" ").toTypedArray()
+        return elementMath
+    }
+
+    private fun postfix(elementMath: Array<String>): Array<String> {
+        var s = ""
+        val elementPostfix: Array<String>
+        val stack = Stack<String>()
+        if (elementMath[0][0] == '-') s = "$s 0"
+        for (i in elementMath.indices) {
+            val c = elementMath[i][0]
+            if (!isOperator(c))
+                s = s + " " + elementMath[i]
+            else {
+                if (c == '(') stack.push(elementMath[i])
+                else {
+                    if (c == ')') {
+                        var c1: Char
+                        do {
+                            c1 = stack.peek()[0]
+                            if (c1 != '(') s = s + " " + stack.peek()
+                            stack.pop()
+                        } while (c1 != '(')
+                    } else {
+                        while (!stack.isEmpty() && priority(stack.peek()[0]) >= priority(c)) {
+                            s = s + " " + stack.peek()
+                            stack.pop()
+                        }
+                        stack.push(elementMath[i])
+                    }
+                }
+            }
+        }
+        while (!stack.isEmpty()) {
```

Comment:
for while if lồng nhau ghê quá. có cách nào khác không nhỉ

Reply:
Để t xem lại thuật toán :D